trace_decoder/src/decoding.rs

```diff
@@ -32,11 +34,120 @@ use crate::{
 
 /// Stores the result of parsing tries. Returns a [TraceParsingError] upon
 /// failure.
-pub type TraceParsingResult<T> = Result<T, TraceParsingError>;
+pub type TraceParsingResult<T> = Result<T, Box<TraceParsingError>>;
+
+/// Represents errors that can occur during the processing of a block trace.
+///
+/// This struct is intended to encapsulate various kinds of errors that might
+/// arise when parsing, validating, or otherwise processing the trace data of
+/// blockchain blocks. It could include issues like malformed trace data,
+/// inconsistencies found during processing, or any other condition that
+/// prevents successful completion of the trace processing task.
+#[derive(Debug)]
+pub struct TraceParsingError {
+    block_num: Option<U256>,
+    block_chain_id: Option<U256>,
+    txn_idx: Option<usize>,
+    addr: Option<Address>,
+    h_addr: Option<H256>,
+    slot: Option<U512>,
+    slot_value: Option<U512>,
+    reason: TraceParsingErrorReason, // The original error type
+}
+
+impl std::fmt::Display for TraceParsingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error processing trace: {}", self.reason)?;
+
+        if let Some(block_num) = self.block_num {
+            write!(f, "\nBlock num: {}", block_num)?;
+        }
+        if let Some(block_chain_id) = self.block_chain_id {
+            write!(f, "\nBlock chain id: {}", block_chain_id)?;
+        }
+        if let Some(txn_idx) = self.txn_idx {
+            write!(f, "\nTxn idx: {}", txn_idx)?;
+        }
+        if let Some(addr) = self.addr {
+            write!(f, "\nAddress: {}", addr)?;
+        }
+        if let Some(h_addr) = self.h_addr {
+            write!(f, "\nHashed address: {}", h_addr)?;
+        }
+        if let Some(slot) = self.slot {
+            write!(f, "\nSlot: {:x}", slot)?;
+        }
+        if let Some(slot_value) = self.slot_value {
+            write!(f, "\nSlot value: {:x}", slot_value)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl std::error::Error for TraceParsingError {}
+
+impl TraceParsingError {
+    /// Function to create a new TraceParsingError with mandatory fields
+    pub(crate) fn new(reason: TraceParsingErrorReason) -> Self {
+        Self {
+            block_num: None,
+            block_chain_id: None,
+            txn_idx: None,
+            addr: None,
+            h_addr: None,
+            slot: None,
+            slot_value: None,
+            reason,
+        }
+    }
+
+    /// Builder method to set block_num
+    pub(crate) fn block_num(&mut self, block_num: U256) -> &mut Self {
+        self.block_num = Some(block_num);
+        self
```

Comment:
We could use a function to get this down to one line, but this is bike shedding and I'm good either way.